# Western Chipmunk Phylogenomics Project
Pipeline describing the analysis pipelines used in Herrera et al. (in prep) "Phylogenomics, reproductive trait evolution, and environmental niche modeling reveals a complex history of cryptic speciation and hybridization in chipmunks".This pipeline outlines raw read processing through variant discovery, phylogenetic analysis of western chipmunks for both ddRAD-seq and whole genome sequence data. Here, we performed enivirnmental nich modeling, analysis of bacular morphology (a key reproductive morphological trait), species tree inference, estimate hybridization with species network inference, and summary statistics (D-statistics).

In each section bellow, I describe the pipeline used in a markdown file (follow links below). Scripts referenced in each section can be found inside the respective folders.

Contact:ndh04c (at) gmail.com

Sections

1. [ddRAD data Processing and ddRAD Analyses](https://github.com/NathanaeldHerrera/Chipmunk-phylogenomics/blob/main/1.%20ddRAD%20Processing%20and%20Analyses/ddRAD_Procesing_%26_Analyses.md)
2. [WGS Raw read Processing, Mapping, and QC](https://github.com/NathanaeldHerrera/Chipmunk-phylogenomics/blob/main/2.%20WGS%20Raw%20read%20Processing%2C%20Mapping%2C%20and%20QC/clean-map-QC.md)
3. [Variant Discovery and FASTA Consensus](https://github.com/NathanaeldHerrera/Chipmunk-phylogenomics/blob/main/3.%20Variant%20Discovery%20and%20Fasta%20Consensus/variant_discovery_fasta_consensus.md)
